osm2streets/src/transform/collapse_short_road.rs

```diff
@@ -83,101 +78,78 @@ impl StreetNetwork {
+                .get_mut(&keep_i)
                 .unwrap()
                 .trim_roads_for_merging
                 .extend(trim_roads_for_merging);
         }
 
         self.remove_road(short);
 
-        // Arbitrarily keep src_i and destroy dst_i. Don't actually remove the intersection until
-        // later; remove_road needs the intersection to exist
-
-        // Fix up all roads connected to dst_i. Delete them and create a new copy; the ID changes,
-        // since one intersection changes.
-        let mut deleted = vec![short];
-        let mut created = Vec::new();
-        let mut old_to_new = BTreeMap::new();
-        let mut new_to_old = BTreeMap::new();
-        for r in self.intersections[&dst_i].roads.clone() {
-            deleted.push(r);
-            let mut road = self.remove_road(r);
-            let old_endpts = RoadWithEndpoints::new(&road);
-
-            let mut new_id = r;
-            if road.src_i == dst_i {
-                new_id.i1 = src_i;
-                road.src_i = src_i;
-            } else {
-                assert_eq!(road.dst_i, dst_i);
-                new_id.i2 = src_i;
-                road.dst_i = src_i;
-            }
-            road.id = new_id;
+        let destroy_i = self.intersections.remove(&destroy_i).unwrap();
 
-            if new_id.i1 == new_id.i2 {
-                // When collapsing many roads around some junction, we wind up with loops. We can
-                // immediately discard those.
-                continue;
-            }
+        // If the intersection types differ, upgrade the surviving interesting.
+        if destroy_i.control == ControlType::TrafficSignal {
+            self.intersections.get_mut(&keep_i).unwrap().control = ControlType::TrafficSignal;
+        }
 
-            old_to_new.insert(r, new_id);
-            new_to_old.insert(new_id, old_endpts);
+        // Remember the merge
+        self.intersections
+            .get_mut(&keep_i)
+            .unwrap()
+            .osm_ids
+            .extend(destroy_i.osm_ids);
 
-            self.insert_road(road);
-            created.push(new_id);
-        }
+        // Fix the endpoint of all roads connected to destroy_i.
+        for r in destroy_i.roads {
+            self.intersections.get_mut(&keep_i).unwrap().roads.push(r);
 
-        // If the intersection types differ, upgrade the surviving interesting.
-        {
-            // Don't use delete_intersection; we're manually fixing up connected roads
-            let i = self.intersections.remove(&dst_i).unwrap();
-            if i.control == ControlType::TrafficSignal {
-                self.intersections.get_mut(&src_i).unwrap().control = ControlType::TrafficSignal;
+            let road = self.roads.get_mut(&r).unwrap();
+            if road.src_i == destroy_i.id {
+                road.src_i = keep_i;
+            } else {
+                assert_eq!(road.dst_i, destroy_i.id);
+                road.dst_i = keep_i;
             }
```

Comment:
Aha, you're right! I forgot about this case; it's hard for me to picture without drawing it out or using the incremental debug steps. I added a comment for now. Since this PR is complex and you have something stacked on it, I vote for taking another round later at cleaning up some of these cases